### PR TITLE
Bug 2023434: Add Purchase Plan Name to Azure Machine Provider Spec

### PR DESCRIPTION
**Pull request by patrickdillon**
Edited: The original PR added support for both Marketplace and Shared Image Gallery images. I have decreased the scope to only Marketplace images, which means we just need a flag as to when to populate the purchase plan. To avoid a boolean, I have added a `string` field `PurchasePlanName` which when present should have the same value as the `SKU`, and would indicate we should create a plan. Using a string avoids using a boolean, and allows us space if there is ever drift in the Microsoft API where purchase plan name != SKU.  

Upstream types: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d2b4e8ed2313adc270a0911d9d0c64e344f34cc4/api/v1beta1/types.go#L316-L398

The OpenShift cluster-api-provider-azure will need to implement logic based off the upstream (I will open a PR soon): https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d2b4e8ed2313adc270a0911d9d0c64e344f34cc4/azure/services/virtualmachines/virtualmachines.go#L259-L287

### CODE REVIEW

**Comment by patrickdillon** [comment]
Looks like `verify` failed because I was using a newer version of `protoc`. Hopefully running the container generated update fixed it.

**Comment by patrickdillon** [comment]
For use case, see openshift/cluster-api-provider-azure#242

File: machine/v1beta1/types_azureprovider.go [hunk 1]
```diff
@@ -174,6 +174,13 @@ type Image struct {
 	Version string `json:"version"`
 	// ResourceID specifies an image to use by ID
 	ResourceID string `json:"resourceID"`
+	// ThirdPartyImage indicates the image is published by a third party publisher and a Plan
+	// will be generated for it.
+	// +optional
+	ThirdPartyImage bool `json:"thirdPartyImage,omitempty"`
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:180]: We avoid booleans in our API because they usually don't age well.

What would this be called if the value was `ThirdParty`?  Is it an `ImageAuthor`? `ImageSource`?  What is the alternative to third party?

File: machine/v1beta1/types_azureprovider.go [hunk 2]
```diff
@@ -174,6 +174,13 @@ type Image struct {
...
+	// SharedGallery specifies an image to use from an Azure Shared Image Gallery
+	// +optional
+	SharedGallery *AzureSharedGalleryImage `json:"sharedGallery,omitempty"`
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:183]: Is this related to the ThirdPartyImage?  Are they specified together or just coincidental in this PR?

Is there a distinction between nil and empty? If there isn't, the configuration API (distinct from workload API) is more discoverable and less prone to error by not being a pointer.

File: machine/v1beta1/types_azureprovider.go [hunk 3]
```diff
@@ -174,6 +174,13 @@ type Image struct {
 	Version string `json:"version"`
 	// ResourceID specifies an image to use by ID
 	ResourceID string `json:"resourceID"`
+	// ThirdPartyImage indicates the image is published by a third party publisher and a Plan
+	// will be generated for it.
+	// +optional
+	ThirdPartyImage bool `json:"thirdPartyImage,omitempty"`
+	// SharedGallery specifies an image to use from an Azure Shared Image Gallery
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:181]: Why would I want to use this?  What image is it overriding if present?

File: machine/v1beta1/types_azureprovider.go [hunk 4]
```diff
@@ -211,6 +218,18 @@ type SecurityProfile struct {
 	EncryptionAtHost *bool `json:"encryptionAtHost,omitempty"`
 }
 
+// AzureSharedGalleryImage defines an image in a Shared Image Gallery to use for VM creation.
+type AzureSharedGalleryImage struct {
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:222]: inside this type, is empty string a valid value that is distinct from "no opinion"?

File: machine/v1beta1/types_azureprovider.go [hunk 5]
```diff
@@ -211,6 +218,18 @@ type SecurityProfile struct {
...
+	// SubscriptionID is the identifier of the subscription that contains the shared image gallery
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:223]: thoughout, links to the azure concepts would be helpful.

File: machine/v1beta1/types_azureprovider.go (hunk 1 above)
Comment (patrickdillon) [reply machine/v1beta1/types_azureprovider.go:180]: `ImageSource` would probably be closest. From https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/928 they define it as:
`Add ThirdPartyImage flag to Marketplace image to indicate an image published by a third party publisher.`

We already have a field called `Publisher` so it is not good to call it `ImagePublisher`, etc

File: machine/v1beta1/types_azureprovider.go (hunk 2 above)
Comment (patrickdillon) [reply machine/v1beta1/types_azureprovider.go:183]: > Is this related to the ThirdPartyImage? Are they specified together or just coincidental in this PR?

`ThirdPartyImage` is only related to Marketplace Images. There are [three types of images in the upstream API.](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d2b4e8ed2313adc270a0911d9d0c64e344f34cc4/api/v1beta1/types.go#L316-L398) Our API has two images types (ID & marketplace) flattened together and this is adding the third as a separate struct. I'm open to other approaches, but I believe we would have to deprecate fields.

Based on our current use case there is no distinction between nil & empty, so I'm fine with removing the pointer.

File: machine/v1beta1/types_azureprovider.go (hunk 4 above)
Comment (patrickdillon) [reply machine/v1beta1/types_azureprovider.go:222]: empty string would only be valid if all the fields were empty. If AzureSharedGallery images are to be used, all fields are required.

File: machine/v1beta1/types_azureprovider.go (hunk 3 above)
Comment (patrickdillon) [reply machine/v1beta1/types_azureprovider.go:181]: If you use this, you are specifying you are using an image from a Shared Gallery (rather than the Marketplace or by specifying an image ID, which is our current approach for installs). 

If you are using a Shared Gallery image, `publisher`, `sku`, & `offer` from the parent struct will be used to populate a required purchase plan. The fields from this struct would be used to create a ref to the image.

**Comment by patrickdillon** [comment]
/hold 

We don't  need to support shared image galleries, so I can remove that and greatly simplify this API bump.

Just need to figure out name/type for the current bool `ThirdPartyImage`.

File: machine/v1beta1/types_azureprovider.go (hunk 1 above)
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:180]: If it not  published by a third party, who is published by?  microsoft, azure, redhat, someone else?

Comment (patrickdillon) [reply machine/v1beta1/types_azureprovider.go:180]: If it's not published by a third party, it's published by Microsoft or one of its cohorts (Canonical/Ubuntu or in a case that is relevant to OpenShift Windows Container the publisher is `WindowsServer`). 

I can't find in-depth documentation on this. [These repo docs](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/topics/custom-images.md#using-azure-marketplace) are consistent with my reading the code that the distinguishing factor is that "third-party images" need a purchase plan. The cluster api provider has a conditional statement `if third-party marketplace image then generate purchase plan`.

So for our particular use case, when the azure cluster-api provider(CAPZ)  is provisioning a new RHCOS node using a marketplace image, the CAPZ needs to include a purchase plan parameter, but for a first-party Windows Container node, a purchase plan must not be included. 

So we expect the CAPZ to use this API to produce results like this: 
**RHCOS image published by Red Hat**
```go
virtualMachine := &compute.VirtualMachine{
		Plan:     &compute.Plan{
			Publisher: to.StringPtr(image.Publisher), // redhat-rhel
			Name:      to.StringPtr(image.SKU),         // ocp-4
			Product:   to.StringPtr(image.Offer),        // ocp-rhcos-worker
		},
		VirtualMachineProperties: &compute.VirtualMachineProperties{
			StorageProfile: &compute.StorageProfile{
				ImageReference: 	imageReference := &compute.ImageReference{
					Publisher: to.StringPtr(vmSpec.Image.Publisher), // redhat-rhel
					Offer:     to.StringPtr(vmSpec.Image.Offer),           // ocp-rhcos-worker
					Sku:       to.StringPtr(vmSpec.Image.SKU),           // ocp-4
					Version:   to.StringPtr(vmSpec.Image.Version),    // 4.9.7
				},
```

vs

**Windows Server image**
```go
virtualMachine := &compute.VirtualMachine{
		Plan:     nil,
		VirtualMachineProperties: &compute.VirtualMachineProperties{
			StorageProfile: &compute.StorageProfile{
				ImageReference: 	imageReference := &compute.ImageReference{
					Publisher: to.StringPtr(vmSpec.Image.Publisher), // MicrosoftWindowsServer
					Offer:     to.StringPtr(vmSpec.Image.Offer),           // WindowsServer
					Sku:       to.StringPtr(vmSpec.Image.SKU),           // datacenter-core-20h2-with-containers-smalldisk
					Version:   to.StringPtr(vmSpec.Image.Version),    // latest
				},
```

**Comment by JoelSpeed** [comment]
This seems ok to me, though while this is convertible to the upstream API should we need to for future projects (eg. `if purchasePlaneName != "" { thirdPartyImage = true }`), this could be confusing from an end user perspective.

@deads2k The change to make this not a bool doesn't really seem to add value to me, since we are saying the value should match the SKU (we are effectively treating it as a bool, rather than an enum). The only way I can see this being advantageous is if, somewhere down the line, the purchase plan name doesn't match the SKU.
I can see the advantage of the bool conversion for other fields we've discussed in other PRs recently (eg AutomaticRestart = true|false to RestartPolicy = RemainStopped|RestartAlways), but this one doesn't seem to add the same value, perhaps I'm missing the nuance on this one.

**Comment by patrickdillon** [comment]
> This seems ok to me, though while this is convertible to the upstream API should we need to for future projects (eg. `if purchasePlaneName != "" { thirdPartyImage = true }`), this could be confusing from an end user perspective.
> 
> @deads2k The change to make this not a bool doesn't really seem to add value to me, since we are saying the value should match the SKU (we are effectively treating it as a bool, rather than an enum). The only way I can see this being advantageous is if, somewhere down the line, the purchase plan name doesn't match the SKU. I can see the advantage of the bool conversion for other fields we've discussed in other PRs recently (eg AutomaticRestart = true|false to RestartPolicy = RemainStopped|RestartAlways), but this one doesn't seem to add the same value, perhaps I'm missing the nuance on this one.

I agree with Joel. I think the boolean for `thirdPartyImage` makes more sense. Potentially we could change it to something like `usePurchasePlan` but I would lean toward aligning with upstream.

File: machine/v1beta1/types_azureprovider.go [hunk 6]
```diff
@@ -174,6 +174,13 @@ type Image struct {
 	Version string `json:"version"`
 	// ResourceID specifies an image to use by ID
 	ResourceID string `json:"resourceID"`
+	// Type identifies the source of the image and related information, such as purchase plans.
+	// Valid values are "ID", "MarketplaceWithPlan", "MarketplaceNoPlan", and omitted, which
+	// behaves as MarketplaceNoPlan if publisher data is supplied, or ID if not.
+	// For more information about purchase plans, see:
+	// https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage#check-the-purchase-plan-information
+	// +optional
+	Type string `json:"type,omitempty"`
```
Comment (JoelSpeed) [inline machine/v1beta1/types_azureprovider.go:183]: I think this works from a description point of view, I would be tempted to add the valid values as constants and make this a typed string, this should then give us constsants to use during the implementation which should make things cleaner

```
type AzureImageType string

const (
  AzureImageTypeID AzureImageType = "ID"
  AzureImageTypeMarketplaceNoPlan AzureImageType = "MarketPlaceNoPlan"
  AzureImageTypeMarketplaceWithPlan AzureImageType = "MarketPlaceWithPlan"
)
```

WDYT?

File: machine/v1beta1/types_azureprovider.go [hunk 7]
```diff
@@ -174,6 +174,13 @@ type Image struct {
 	Version string `json:"version"`
 	// ResourceID specifies an image to use by ID
 	ResourceID string `json:"resourceID"`
+	// Type identifies the source of the image and related information, such as purchase plans.
+	// Valid values are "ID", "MarketplaceWithPlan", "MarketplaceNoPlan", and omitted, which
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:178]: I suggest reserving the right to change the meaning of `""` in this documentation.  Maybe something like 

Valid values are "ID"< "MarketplaceWithPlan", "MarketplaceNoPlan", and "".
"" means no opinion and the platform chooses a good default which may change over time. Currently that default is "MarketplaceNoPlan" if publisher data is supplied, or "ID" if not.
For more information***

**Review by JoelSpeed — COMMENTED** [review]
/lgtm
